src/fcntl/statx.rs

```diff
@@ -0,0 +1,371 @@
+
+    /// Retrieve file birth (creation) time, if it has been returned by kernel
+    pub fn btime(&self) -> Option<libc::statx_timestamp> {
+        if Mask::STATX_BTIME.bits() & self.inner.stx_mask > 0 {
+            Some(self.inner.stx_btime)
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve inode number, if it has been returned by kernel
+    pub fn ino(&self) -> Option<u64> {
+        if Mask::STATX_INO.bits() & self.inner.stx_mask > 0 {
+            Some(self.inner.stx_ino)
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve file size, in bytes, if it has been returned by kernel
+    pub fn size(&self) -> Option<u64> {
+        if Mask::STATX_SIZE.bits() & self.inner.stx_mask > 0 {
+            Some(self.inner.stx_size)
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve file size as a number of blocks, if it has been returned by kernel
+    pub fn blocks(&self) -> Option<u64> {
+        if Mask::STATX_BLOCKS.bits() & self.inner.stx_mask > 0 {
+            Some(self.inner.stx_blocks)
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve file size as a number of blocks, if it has been returned by kernel
+    pub fn blksize(&self) -> Option<u64> {
+        // I'm not sure what exact mask bit should be used to check presence of block size.
+        // Actual Linux kernel seems return most of STATX_BASIC_STATS in one go, regarless of which things are asked for.
+        if Mask::STATX_BASIC_STATS.bits() & self.inner.stx_mask == Mask::STATX_BASIC_STATS.bits()  {
+            Some(self.inner.stx_blocks)
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve mount ID, if it has been returned by kernel
+    pub fn mnt_id(&self) -> Option<u64> {
+        if Mask::STATX_MNT_ID.bits() & self.inner.stx_mask > 0 {
+            Some(self.inner.stx_mnt_id)
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve device major and minor numbers (first and second elements of the tuple respectively) of the filesystem where the file resides at, if it has been returned by kernel.
+    pub fn dev_major_minor(&self) -> Option<(u32, u32)> {
+        // I'm not sure what exact mask bit should be used to check presence of this information.
+        // Actual Linux kernel seems return most of STATX_BASIC_STATS in one go, regarless of which things are asked for.
+        if Mask::STATX_BASIC_STATS.bits() & self.inner.stx_mask == Mask::STATX_BASIC_STATS.bits() {
+            Some((self.inner.stx_dev_major, self.inner.stx_dev_minor))
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve pointed-to device major and minor numbers (first and second elements of the tuple respectively) of this character or block device file, if it has been returned by kernel.
+    /// Note that this function does not check for the file type. It would return `Some` even for regular files.
+    pub fn rdev_major_minor(&self) -> Option<(u32, u32)> {
```

Comment:
What's the difference compared to the previous method?

Reply:
For example, `/dev/null` character special file lives in a `/dev/` directory of mounted filesystem addressed by a device number `5h/5d`. But `/dev/null` itself _points_ to another, non-filesystem device of type `1,3`. If you `cp -a /dev/null /root/null`, the resulting file would have different `dev_major_minor` (as it is now on another filesystem), but would point to the same device (`1,3`), so the same `rdev_major_minor`. `dev_major_minor` is relevant for all fiels on Linux, `rdev_major_minor` is relevant only for block and character special files. `dev_major_minor` + `ino` = sort of a numeric identifier of a specific file on Linux.

Reply:
But, as usual, there may be copy&paste sloppiness here, so implementation may return wrong number.

How such syscall wrappers are supposed to be tested? Shall unit tests create some tricky files (as root) to be able to properly test all those things? Making Linux kernel to return some newer attributes may be nontrivial.

Reply:
Ok, it must be a copy/paste error then.  Please fix it.

As for testing, you can add tests in the test directory.  You don't need to worry about testing every feature advertised by the kernel; that's a job for the kernel test suite.  You just need to test enough features to prove that the Rust bindings work.